### ISSUE TITLE
Support HiDPI (Retina) devices

### ISSUE THREAD

**Issue by chearon** [richardbutler/node-spritesheet#9]
I've been thinking about it, and I think the nicest way to do it would be to define the device-pixel ratio of your PNGs, and then define the desired output ratios - say I have 16x16 spans, and 32x32 PNGs.  I envision:

```
options: {
  outputImage: 'images/spritesheets/flags.png',
  outputCss: 'stylesheets/flags.css',
  selector: '.flag',
  inputDevicePixelRatio: 2,
  outputDevicePixelRatios: [1, 1.5, 2]
}
```

would generate a stylesheet with `@media` queries for each devicePixelRatio, and all with `background-size: 16px 16px`, which comes from `image.width / inputDevicePixelRatio` and `image.height / inputDevicePixelRatio`.  The images (or perhaps the tiled image, rather than the individual images?) would be scaled down by ImageMagick via:

```
  width: image.width / inputDevicePixelRatio * outputDevicePixelRatio
  height: image.height / inputDevicePixelRatio * outputDevicePixelRatio
```

Actually, there'd be no `@media` query or `background-size` for d-p ratios of 1 for compatibility with old browsers, which would also mean that spritesheet produces the exact same output it used to with old config files.  I might be able to implement this but it involves significant changes to the Layout, Style and Builder code, so, what do you think @richardbutler?

**Comment by richardbutler** [comment]
Good point, this isn't a feature I'd actually considered in this sense, but I like the idea.

My only consideration would be which interpolation methods are available with ImageMagick, as I would want it to be comparable with Photoshop's bicubic sharper. For instance, nearest neighbour would be a no-go, as would anything else that would blur the result. I'll do some digging around.

Either way, it could just as easily have an option (or default to) looking for "@2x" versions of each image.

**Comment by chearon** [comment]
OK, hopefully ImageMagick is flexible enough to do it.  Having spritesheet look for alternate resolutions already provided by the designer would be a perfect solution for those who prefer Photoshop.  Maybe look for a suffix of the filename without the extension ("_2x"), which can be overridden with a string or a function to call in case they're already using a different naming convention.

Anyway in my case I think the downsizing my ImageMagick is good enough.  Keep me posted on what you find.  I may start hacking on this soon cause I might need it.

By the way, what's the algorithm in Layout.coffee?

**Comment by richardbutler** [comment]
Will do - I'm intrigued by this now, so I may well have a poke around when I get some time. It seems ImageMagick supports various interpolations, bicubic being one of them, so hopefully it'll be close enough. I'll test them and see what I find.

The layout uses a [bin packing algorithm](http://codeincomplete.com/posts/2011/5/7/bin_packing) - I can't lay claim to it though, it's mostly [Mr Gordon's](https://github.com/jakesgordon/sprite-factory), as is most of the clever stuff.

**Comment by chearon** [comment]
Cool, sounds good - I'll stand by.  Definitely just added that blog post to my reading list.

**Comment by richardbutler** [comment]
It's not the cleanest-ever code, as it's quite a complex problem, but I had a bit of time to fart-arse about with it, so there's a version on the [retina branch](https://github.com/richardbutler/node-spritesheet/tree/retina).

For now, it just assumes you've supplied the retina versions and it downsamples. The tricky thing with sniffing for a "2x" with the Grunt task is that you'd have to segregate your retina files from your regular files by way of different directories due to the way files are referenced with minimatch.

I've tested a couple of scenarios with it, and it seems to work well, but shout if there's anything I've missed.

**Comment by chearon** [comment]
Nice, it's working well for me.  I know it's more "pure" to maintain your own separate resolution sets, but it's really useful to me to only have to work on the high resolution set, and have Spritesheet downsize them for me.  Perhaps this could be optional functionality?

Before you posted this I implemented [CSS selector customization per-image](https://github.com/jakesgordon/sprite-factory#customizing-the-css-selector-per-image).  Since your port treats the `selector` option as a rule on the actual image, rather than a selector "prefix" as it is in Sprite Factory, I made `selector` apply only to the last selector that results from the image name.  So it shouldn't break any config files already out there.

I had to patch Builder to use `path` for this because it was breaking on filenames with multiple dots like `div.icon-outer--hover__gb.png`.

**Comment by richardbutler** [comment]
Yeah, Sprite Factory is a little more fully-featured, I just ported over the bits I needed as a basis. I'm not all that sure I like the look of that selector feature, it looks a little clumsy to me. I can see the merit of wanting to derive a selector from a filename, but to actually put CSS there seems too closely-coupled.

Regardless, it would probably already be achievable by using the [resolveImageSelector](https://github.com/richardbutler/node-spritesheet/tree/retina#complex-example) function to augment the filename.

**Comment by richardbutler** [comment]
Retina feature branch merged: 1765ec44ed5cc72a7c58c0ea14f809175d584855.

**Comment by chearon** [comment]
I agree that `resolveImageSelector` is a better solution, however, you're still applying `selector` to the front of whatever selector the user defines.  That means I can't really specify an outer selector if I want to.  For example, I want all of my PNGs that end in `-hover` to be active when an outside div is hovered-over.  Like the way the "Watching thread" button below changes the blue to white when you hover over the text.

The code might have looked a bit clumsy but was error-proof, backwards compatible, and more flexible than this.
